README.rst
===============================
HydroFunctions
===============================

.. image:: https://img.shields.io/pypi/v/hydrofunctions.svg
        :target: https://pypi.python.org/pypi/hydrofunctions
        :alt: Visit Hydrofunctions on PyPI

.. image:: https://img.shields.io/travis/mroberge/hydrofunctions.svg
        :target: https://travis-ci.org/mroberge/hydrofunctions
        :alt: Unit Testing Status

.. image:: https://codecov.io/gh/mroberge/hydrofunctions/branch/master/graph/badge.svg
        :target: https://codecov.io/gh/mroberge/hydrofunctions
        :alt: Code Coverage Status

.. image:: https://readthedocs.org/projects/hydrofunctions/badge/?version=latest
        :target: https://hydrofunctions.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/github/license/mashape/apistatus.svg
        :target: https://github.com/mroberge/hydrofunctions/blob/master/LICENSE
        :alt: MIT license

A suite of convenience functions for exploring water data in Python.

Features
--------

* Retrieves stream data from the USGS NWIS service
* Select data using multiple site numbers, by state, county codes, or a boundary box
* Preserves NWIS metadata, including NoData values
* Helpful error messages to help you write valid requests
* Extracts data into a Pandas dataframe, json, or dict
* Plot beautiful graphs in Jupyter Notebooks
   * hydrographs (or time series of any data)
   * flow duration charts
   * cycle plots to illustrate annual or diurnal cycles
   * Interactive map for finding stream gauge ID numbers
* Plotting and manipulation through Pandas dataframes
* Retrieve rating curves and field notes for sites
* Saves data in compact, easy-to-use parquet files instead of requesting the same dataset repeatedly
* Still in active development! Let me know what features you want!

Read the `Users Guide <https://hydrofunctions.readthedocs.io/en/master>`_ for more details.


Basic Usage
-----------

First, import hydrofunctions into your project. If you plan to work with Jupyter
notebooks, then go ahead and enable automatic chart display:

.. code-block:: ipython

    In  [1]: import hydrofunctions as hf
             %matplotlib inline

Create an NWIS data object to hold our request and the data we will retrieve.
We will request the instantaneous values ('iv') for site '01585200' for the
past 55 days:

.. code-block:: ipython

    In  [2]: herring = hf.NWIS('01585200', 'iv', period='P55D')
    Requested data from https://waterservices.usgs.gov/nwis/iv/?format=json%2C1.1&sites=01585200&period=P55D

Check that the request went smoothly:

.. code-block:: ipython

    In  [3]: herring.ok
    Out [3]: True

Find out what data we received:

.. code-block:: ipython

    In  [4]: herring
    Out [4]: USGS:01585200: WEST BRANCH HERRING RUN AT IDLEWYLDE, MD
                 00060: <5 * Minutes>  Discharge, cubic feet per second
                 00065: <5 * Minutes>  Gage height, feet
             Start: 2019-05-25 01:05:00+00:00
             End:   2019-07-19 19:05:00+00:00

This tells us the name of our site, and gives a list of the parameters that we
have. For each parameter it lists how frequently the data were collected, and
it show the common name of the parameter and its units.

Create a dataframe using only our discharge data, and list the first five items:

.. code-block:: ipython

    In  [5]: herring.df('discharge').head()
    Out [5]:

*--a table with our data appears--*

    +------------------------------+---------------------------+
    |          datetimeUTC         | USGS:01585200:00060:00000 |
    +------------------------------+---------------------------+
    |   2019-05-25 01:05:00+00:00  |                1.57       |
    +------------------------------+---------------------------+
    |   2019-05-25 01:10:00+00:00  |                1.57       |
    +------------------------------+---------------------------+
    |   2019-05-25 01:15:00+00:00  |                1.51       |
    +------------------------------+---------------------------+
    |   2019-05-25 01:20:00+00:00  |                1.57       |
    +------------------------------+---------------------------+
    |   2019-05-25 01:25:00+00:00  |                1.57       |
    +------------------------------+---------------------------+

Plot the data using built-in methods from Pandas and mathplotlib:

.. code-block:: ipython

    In  [6]: herring.df('q').plot()
    Out [6]:

*--a stream hydrograph appears--*

.. image:: _static/HerringHydrograph.png
   :alt: a stream hydrograph for Herring Run

Draw an interactive map in a Jupyter Notebook:

.. code-block:: ipython

    In  [7]: hf.draw_map()
    Out [7]:

.. image:: _static/draw_map.jpg
        :alt: a map in an interactive Jupyter Notebook.

Learn more:

* `More usage <https://hydrofunctions.readthedocs.io/en/master/usage.html>`_ tips
* `Introduction to Hydrofunctions <https://github.com/mroberge/hydrofunctions/blob/master/notebooks/Introduction%20to%20Hydrofunctions.ipynb>`_, a Jupyter Notebook with a quick tutorial.

Easy Installation
-----------------

The easiest way to install Hydrofunctions is by typing this from your
command line:

.. code-block:: console

    $ pip install hydrofunctions


Hydrofunctions depends upon Pandas and numerous other scientific packages
for Python. `Anaconda <https://www.continuum.io/open-source-core-modern-software>`_
is an easy, safe, open-source method for downloading everything and avoiding
conflicts with other versions of Python that might be running on your
computer.

Visit the `Installation Page <https://hydrofunctions.readthedocs.io/en/master/installation.html>`_
in the Users Guide to learn how to install
Anaconda, or if you have problems using the Easy Installation method above.


Other Projects You Should See
-----------------------------

`Hydropy <https://github.com/stijnvanhoey/hydropy>`_, a Python package that builds upon Pandas for enhanced data selection and plotting of hydrology data.

`WellApplication <https://github.com/inkenbrandt/WellApplication>`_ a Python package that provides functions for working with dataloggers and USGS well data.

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage

MIT License

Copyright (c) 2016, Martin Roberge and contributors


HISTORY.rst
=======
History
=======

0.1.0 (2016-08-11)
------------------

* Project set up.

0.1.1 (2016-08-11)
------------------

* Customized project boilerplate.

0.1.2 (2016-08-26)
------------------

* Set up template for custom exceptions
* Added get_nwis function

0.1.3 (2016-09-09)
-----------------------

* Check user inputs & raise explanatory exceptions
* Extract data from response into a dataframe.
* Stations object for managing data.

0.1.4 (2016-09-18)
----------------------

* Added tests & documentation.

0.1.5 (2018-02-22)
----------------------

* Updated to support Python 3.6
* Updated docs, added notebooks (mcr jdh)
* Added parameterCd to allow requests for different datasets (thanks @jdhughes-usgs!)
* Added ability to query sites by state or county (jdh)
* Added ability to request lists of sites or counties (jdh)
* Improved column names: now includes site id & variable description (for example, '07228000 - Mean Discharge, cubic feet per second')(jdh)
* Added descriptive warnings to explain why queries fail (mcr)


0.1.6 (2018-03-07)
----------------------

* Added draw_map function for help selecting sites (mcr)
* Added ability to query sites by a bounding box (jdh)
* Revised pandas dataframe column names to be the name composed of the provider, site id, parameter, and statistic (for example. USGS:01638500:00060:0001). (jdh)
* Added qualifiers for each station as a column in the dataframe. (jdh)
* Added ability extract NWIS property data from the response object (for example, siteName). (jdh)

0.1.7 (2018-11-11)
----------------------

* Added a flow duration chart (mcr)
* Added the cycleplot chart (mcr)
* Added code coverage (mcr)
* Improved the build scripts for TravisCI (mcr)
* Updated to support Python 3.7

0.1.8 (in development)
----------------------

* Dropped Python 3.4 support.
* parameterCd now accepts multiple parameters in request.
* If parameterCd is not specified, then all available parameters will be requested (default).
* Improved parsing of data from NWIS:
    - missing observations are noted
    - duplicates found & removed
    - unsorted data found & cleaned.
    - different frequencies raise a warning when resampled
* Missing data can now be replaced with interpolated values.
* hf.rating_curve(site) returns the current rating curve for a site.
* hf.field_meas(site) returns the field data and notes used to create a rating curve.
* NWIS.df() allows you to create dataframes with subsets of the data, such as only 'discharge' or 'stage' or data quality 'flags'.
* New and improved REPR: lists stations & parameters in dataset.
* Saving data to a file:
    - the 'file' parameter for NWIS allows you to save your data locally
    - If the file doesn't exist, NWIS requests the data and creates the file
    - Uses the parquet format for faster load times and smaller file sizes




hydrofunctions/station.py
# -*- coding: utf-8 -*-

"""
hydrofunctions.station
~~~~~~~~~~~~~~~~~~~~~~

This module contains the Station and NWIS classes, which are used for
organizing and managing data for data collection sites.

-----
"""
from __future__ import absolute_import, print_function, division, unicode_literals
import re

import json
from . import typing
from . import hydrofunctions as hf
from . import helpers


class Station(object):
    """A class for organizing stream gauge data for a single request.
    """
    station_dict = {}

    def __init__(self, site=None):
        Station.station_dict[site] = self
        self.site = site
        # One option is to make it so that you can pass in a get_data function
        # during the creation of an instance.
        self.get_data = None


class NWIS(Station):
    """A class for working with data from the USGS NWIS service.

    description

    Args:
        site (str or list of strings):
            a valid site is '01585200' or ['01585200', '01646502']. Default is
            None. If site is not specified, you will need to select sites using
            stateCd or countyCd.

        service (str):
            can either be 'iv' or 'dv' for instantaneous or daily data.
            'dv'(default): daily values. Mean value for an entire day.
            'iv': instantaneous value measured at this time. Also known
            as 'Real-time data'. Can be measured as often as every
            five minutes by the USGS. 15 minutes is more typical.

        start_date (str):
           should take on the form 'yyyy-mm-dd'

        end_date (str):
            should take on the form 'yyyy-mm-dd'

        stateCd (str):
            a valid two-letter state postal abbreviation, such as 'MD'. Default
            is None. Selects all stations in this state. Because this type of
            site selection returns a large number of sites, you should limit
            the amount of data requested for each site.

        countyCd (str or list of strings):
            a valid county FIPS code. Default is None. Requests all stations
            within the county or list of counties. See https://en.wikipedia.org/wiki/FIPS_county_code
            for an explanation of FIPS codes.

        bBox (str, list, or tuple):
            a set of coordinates that defines a bounding box.
                * Coordinates are in decimal degrees.
                * Longitude values are negative (west of the prime meridian).
                * Latitude values are positive (north of the equator).
                * comma-delimited, no spaces, if provided as a string.
                * The order of the boundaries should be: "West,South,East,North"
                * Example: "-83.000000,36.500000,-81.000000,38.500000"

        parameterCd (str or list of strings):
            NWIS parameter code. Usually a five digit code. Default is 'all'.
            A valid code can also be given as a list: parameterCd=['00060','00065']
            This will request data for this parameter.

                * if value is 'all', or no value is submitted, then NWIS will \
                return every parameter collected at this site. (default option)
                * stage: '00065'
                * discharge: '00060'
                * not all sites collect all parameters!
                * See https://nwis.waterdata.usgs.gov/usa/nwis/pmcodes for full list

        period (str):
            NWIS period code. Default is None.
                * Format is "PxxD", where xx is the number of days before \
                today, with a maximum of 999 days accepted.
                * Either use start_date or period, but not both.


    """

    def __init__(self,
                 site=None,
                 service='dv',
                 start_date=None,
                 end_date=None,
                 stateCd=None,
                 countyCd=None,
                 bBox=None,
                 parameterCd='all',
                 period=None,
                 file=None):

        self.ok = False
        if file is not None:
            try:
                self._dataframe, self.meta = hf.read_parquet(file)
                self.ok = True
                print('Reading data from', file)

            except OSError as err:
                # File does not exist yet, we'll make it later.
                pass

        if self.ok == False:
            self.response = hf.get_nwis(site,
                                        service,
                                        start_date,
                                        end_date,
                                        stateCd=stateCd,
                                        countyCd=countyCd,
                                        bBox=bBox,
                                        parameterCd=parameterCd,
                                        period=period
                                        )
            try:
                self.json = self.response.json()
                self._dataframe, self.meta = hf.extract_nwis_df(self.json)
                self.ok = self.response.ok
                if file is not None:
                    self.save(file)
                    print('Saving data to', file)
            except json.JSONDecodeError as err:
                self.ok = False
                print('JSON decoding error. URL: {self.response.url}')
                raise json.JSONDecodeError(err)

        # Can I get rid of this, and only keep metadata in the meta dict?
        if self.ok:
            self.site = site
            self.service = service
            self.start_date = start_date
            self.end_date = end_date
            self.start = self._dataframe.index.min()
            self.end = self._dataframe.index.max()

    def __repr__(self):
        repr_string = ""
        for site_id in sorted(self.meta.keys()):
            repr_string += site_id + ": " + self.meta[site_id]['siteName'] + "\n"
            for param in sorted(self.meta[site_id]['timeSeries'].keys()):
                repr_string += "    " + param + ": " + \
                    self.meta[site_id]['timeSeries'][param]['variableFreq'] + \
                    "  " + self.meta[site_id]['timeSeries'][param]['variableDescription'] + "\n"
        repr_string += "Start: " + str(self.start) + "\n" + \
        "End:   " + str(self.end)
        return repr_string

    def df(self, *args):
        """
        Return a subset of columns from the dataframe.

        Args:
            '': If no args are provided, the entire dataframe will be returned.

            str 'all': the entire dataframe will be returned.

            str 'data': all of the parameters will be returned, with no flags.

            str 'flags': Only the _qualifier flags will be returned. Unless the \
            flags arg is provided, only data columns will be returned. Visit \
            https://waterdata.usgs.gov/usa/nwis/uv?codes_help#dv_cd1 to see a \
            more complete listing of possible codes.

            str 'discharge' or 'q': discharge columns ('00060') will be returned.

            str 'stage': Gauge height columns ('00065') will be returned.

            int any five digit number: any matching parameter columns will be returned. '00065' returns stage, for example.

            int any eight to twelve digit number: any matching stations will be returned.
        """
        data_cols = self._dataframe.columns.str.contains(r'[0-9]$') # Data ends in a number.
        flag_cols = self._dataframe.columns.str.contains('_qualifiers')
        Q_cols = self._dataframe.columns.str.contains(':00060:') # This includes data & flags
        stage_cols = self._dataframe.columns.str.contains(':00065:')
        all_cols = self._dataframe.columns != ""
        param_re = r'^\d{5}$' # parameters are a five-digit number.
        station_re = r'\d{8,12}$' # station ID's are between 8 and 12 digits.

        sites = all_cols
        params = all_cols
        meta = all_cols
        if len(args) == 0:
            pass
        else:
            meta = data_cols
            for item in args:
                if item == 'all':
                    sites = all_cols
                    params = all_cols
                    meta = all_cols
                    break
                elif item == 'discharge':
                    params = Q_cols
                elif item == 'q':
                    params = Q_cols
                elif item == 'stage':
                    params = stage_cols
                elif item == 'data':
                    meta = data_cols
                elif item == 'flags':
                    meta = flag_cols
                elif re.search(param_re, item):
                    param_arg = ":" + item + ":"
                    params = self._dataframe.columns.str.contains(param_arg)
                    if not params.any():
                        raise ValueError("The parameter '{param}' is not contained in this dataset.".format(param=item))
                elif re.search(station_re, item):
                    station_arg = ":" + item + ":"
                    sites = self._dataframe.columns.str.contains(station_arg)
                    if not sites.any():
                        raise ValueError("The site '{site}' is not in this dataset.".format(site=item))
                else:
                    raise ValueError("The argument '{item}' is not recognized.".format(item=item))
        selection = sites & params & meta
        requested_df = self._dataframe.loc[:, selection]
        return requested_df

    def get_data(self):
        print("It is no longer necessary to call .get_data() to request data.")
        return self

    def save(self, file):
        hf.save_parquet(file, self._dataframe, self.meta)
        return self

    def read(self, file):
        self._dataframe, self.meta = hf.read_parquet(file)
        return self
